engine/kv_engine_cleaner.cpp

```diff
@@ -235,366 +237,432 @@ void KVEngine::cleanNoHashIndexedSkiplist(
-                  bool success =
-                      Skiplist::Remove(dl_record, node, pmem_allocator_.get(),
-                                       skiplist_locks_.get());
-                  kvdk_assert(success, "");
-                  hash_table_->Erase(&(*slot_iter));
-                  purge_dl_records.emplace_back(dl_record);
-                  need_purge_num++;
-                }
-                break;
+              break;
+            }
+            case PointerType::Skiplist: {
+              Skiplist* skiplist = slot_iter->GetIndex().skiplist;
+              total_num += skiplist->Size();
+              auto head_record = skiplist->HeaderRecord();
+              auto old_record =
+                  removeOutDatedVersion<DLRecord>(head_record, min_snapshot_ts);
+              need_purge_num +=
+                  PushToOutdatedPool(old_record, purge_dl_records);
+              if ((slot_iter->GetRecordStatus() == RecordStatus::Outdated ||
+                   head_record->GetExpireTime() <= now) &&
+                  head_record->entry.meta.timestamp < min_snapshot_ts) {
+                hash_table_->Erase(&(*slot_iter));
+                pending_clean_records.outdated_skip_lists.emplace_back(
+                    std::make_pair(version_controller_.GetCurrentTimestamp(),
+                                   skiplist));
+                need_purge_num += skiplist->Size();
+              } else if (!skiplist->IndexWithHashtable()) {
+                pending_clean_records.no_index_skiplists.emplace_back(skiplist);
               }
-              case PointerType::DLRecord: {
-                total_num++;
-                auto dl_record = slot_iter->GetIndex().dl_record;
-                auto old_record =
-                    removeOutDatedVersion<DLRecord>(dl_record, min_snapshot_ts);
-                if (old_record) {
-                  purge_dl_records.emplace_back(old_record);
-                  need_purge_num++;
-                }
-                if (slot_iter->GetRecordStatus() == RecordStatus::Outdated &&
-                    dl_record->entry.meta.timestamp < min_snapshot_ts) {
-                  bool success = Skiplist::Remove(dl_record, nullptr,
-                                                  pmem_allocator_.get(),
-                                                  skiplist_locks_.get());
-                  kvdk_assert(success, "");
-                  hash_table_->Erase(&(*slot_iter));
-                  purge_dl_records.emplace_back(dl_record);
-                  need_purge_num++;
-                }
-                break;
+              break;
+            }
+            case PointerType::List: {
+              List* list = slot_iter->GetIndex().list;
+              total_num += list->Size();
+              auto current_ts = version_controller_.GetCurrentTimestamp();
+              auto old_list = removeListOutDatedVersion(list, min_snapshot_ts);
+              if (old_list) {
```

Comment:
Always unlink first, then get timestamp. The timestamp functions as a barrier.
List removed | ts | visit list, cannot visit removed list, and removed list can be safely destroyed.
ts | remove List
ts | visit List, may visit a destroyed List

Reply:
updated